nextest-runner/src/platform.rs

```diff
@@ -5,64 +5,374 @@
+    /// Returns an error if the host platform could not be determined.
+    pub fn set_target(&mut self, triple: TargetTriple) -> BuildPlatformTargetBuilder<'_> {
+        BuildPlatformTargetBuilder {
+            libdir: None,
+            parent: self,
+            triple,
+        }
+    }
+
+    /// Set the host libdir to the built [`BuildPlatform`].
     ///
+    /// Used to set the dynamic linker search path when running the test executables. If we fail to
+    /// parse the input, the libdir won't be set.
+    pub fn set_host_libdir(&mut self, reader: impl io::BufRead) {
+        self.host_libdir = read_first_line_as_path(reader);
+    }
+
+    /// Creates a new [`BuildPlatform`].
+    ///
+    /// # Error
     /// Returns an error if the host platform could not be determined.
-    pub fn new(target: Option<TargetTriple>) -> Result<Self, UnknownHostPlatform> {
+    pub fn build(self) -> Result<BuildPlatform, UnknownHostPlatform> {
         let host = Platform::current().map_err(|error| UnknownHostPlatform { error })?;
-        Ok(Self { host, target })
+        Ok(BuildPlatform {
+            host,
+            host_libdir: self.host_libdir,
+            target: self.target,
+        })
     }
+}
 
-    /// Creates a new `BuildPlatform` where the host is specified.
+/// Populate a [`BuildPlatformTarget`] in the [`BuildPlatform`] built by
+/// [`BuildPlatformBuilder`].
+pub struct BuildPlatformTargetBuilder<'a> {
+    parent: &'a mut BuildPlatformBuilder,
+    triple: TargetTriple,
+    libdir: Option<Utf8PathBuf>,
+}
+
+impl<'a> BuildPlatformTargetBuilder<'a> {
+    /// Add the libdir to the target.
     ///
-    /// This is intended for testing situations. Most users should call [`Self::new`] instead.
-    pub fn new_with_host(host: Platform, target: Option<TargetTriple>) -> Self {
-        Self { host, target }
+    /// Used to set the dynamic linker search path when running the test executables. If we fail to
+    /// parse the input, the libdir won't be set.
+    pub fn set_libdir(&mut self, reader: impl io::BufRead) {
+        self.libdir = read_first_line_as_path(reader);
     }
 
+    /// Build a [`BuildPlatform`] and add it to the [`BuildPlatformBuilder`].
+    pub fn add(self) {
+        self.parent.target = Some(BuildPlatformTarget {
+            triple: self.triple,
+            libdir: self.libdir,
+        });
+    }
+}
+
+/// The target platform.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct BuildPlatformTarget {
+    /// The target triplet, which consists of machine, vendor and OS.
+    pub triple: TargetTriple,
+
+    /// The target libdir.
+    pub libdir: Option<Utf8PathBuf>,
+}
+
+/// A representation of host and target platform.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct BuildPlatform {
+    /// The host platform.
+    pub host: Platform,
+
+    /// The host libdir.
+    pub host_libdir: Option<Utf8PathBuf>,
+
+    /// The target platform, if specified.
+    pub target: Option<BuildPlatformTarget>,
```

Comment:
@sunshowers This design makes the builder complicate, e.g.:

```rust
        let mut builder = BuildPlatformBuilder::default();
        let mut target_builder = builder.set_target(TargetTriple::x86_64_unknown_linux_gnu());
        target_builder.set_libdir(Cursor::new("/fake/target/libdir"));
        target_builder.add();
        let build_platform = builder
            .build()
            .expect("should build with target successfully");
```

The alternative design is a flat struct
```rust
pub struct BuildPlatform {
    pub host: Platform,
    pub host_libdir: Option<Utf8PathBuf>,
    pub target_libdir: Option<Utf8PathBuf>,
    pub target_triple: Option<TargetTriple>,
}
```
and the previous build code can be rewritten as
```rust
        let mut builder = BuildPlatformBuilder::default();
        builder
            .set_target_triple(TargetTriple::x86_64_unknown_linux_gnu())
            .set_target_libdir(Cursor::new("/fake/target/libdir"));
        let build_platform = builder
            .build()
            .expect("should build with target successfully");
```

Which way do you think is better?

Reply:
Well -- the flat struct, as your proposed it, allows for the invalid state where `target_libdir` is `Some(_)` but `target_triple` is `None`. So I'd strongly prefer the `BuildPlatformTarget` pattern you have.

I'm not quite sure the builder pattern is pulling its weight though. Can the setters just live directly on the `BuildPlatforms` and `BuildPlatformTarget` structs?